### PR TITLE
Further refactored APIs to be cleaner + Added better image control

### PR DESCRIPTION
**Pull request by Haud**
Major refactor:

* Renamed `CardNumberTextField` to `CardTextField` to differentiate it better from `NumberTextField`
* simplified APIs and logic across the board
* Reduced state in the card text field by reducing a lot of properties and reduced optionals
* Added `CardTypeImageStore` protocol to make it easier to specify custom images by implementers

@michaelCampbell & @DannyVancura

### CODE REVIEW

File: Pod/Classes/UITextField+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//  Created by Christopher Jones on 3/24/16.
+//
+//
+
+import UIKit
+
+internal extension UITextField {
+
+    internal static var emptyTextFieldCharacter: String {
+        return "\u{202F}"
```
Comment (MichaelCampbell) [inline Pod/Classes/UITextField+Extensions.swift]: what is this dark magic?

Comment (Haud) [inline Pod/Classes/UITextField+Extensions.swift]: I think it's backspace.

Comment (DannyVancura) [inline Pod/Classes/UITextField+Extensions.swift]: It's a very short empty text character. Methods like `textField: shouldChangeCharactersInRange:` or `deleteBackward` aren't called if the text field is empty (the latter is called - but only when entered from a hardware keyboard). Therefor, this character is inserted and as soon as the user hits backspace, the character gets deleted - text.characters.count becomes 0 -> we can go to the previous text field.
PaymentKit used a slightly different (even shorter) character, which caused weird behavior when used in a right-aligned text field: The cursor jumped to the left when starting editing.

tl;dr:
It's a barely visible character for detecting "backspace actions".

Comment (Haud) [inline Pod/Classes/UITextField+Extensions.swift]: clever solution

Comment (MichaelCampbell) [inline Pod/Classes/UITextField+Extensions.swift]: wow, that's interesting

Comment (prolificphotis) [inline Pod/Classes/UITextField+Extensions.swift]: Could we put that entire explanation into a comment above Daniel?

On Thu, Mar 24, 2016 at 2:30 PM, Michael Campbell <notifications@github.com>
wrote:

> In Pod/Classes/UITextField+Extensions.swift
> <https://github.com/prolificinteractive/Caishen/pull/10#discussion_r57391096>
> :
>
> > @@ -0,0 +1,18 @@
> > +//
> > +//  UITextField+Extensions.swift
> > +//  Pods
> > +//
> > +//  Created by Christopher Jones on 3/24/16.
> > +//
> > +//
> > +
> > +import UIKit
> > +
> > +internal extension UITextField {
> > +
> > +    internal static var emptyTextFieldCharacter: String {
> > +        return "\u{202F}"
>
> wow, that's interesting
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/prolificinteractive/Caishen/pull/10/files/ea668a9b8b4b5035bfd4e1f0fe623d4e71911fd9#r57391096>
>



-- 

Photis Patriotis

VP of Engineering

Prolific Interactive

BK // 45 Main Street, Suite 1006 | Brooklyn, NY 11201

SF // 995 Market Street, Suite 1401 | San Francisco, CA 94103

m: 215.480.9759 | prolificinteractive.com | @weareprolific
<http://www.twitter.com/weareprolific>

Comment (Haud) [inline Pod/Classes/UITextField+Extensions.swift]: +1

Comment (DannyVancura) [inline Pod/Classes/UITextField+Extensions.swift]: Makes sense! I added it to [my pull request](https://github.com/prolificinteractive/Caishen/pull/13) for the `refactor` branch.
